### PR TITLE
(Feature)|Add customParameters to OAuth2AuthorizationResponse

### PR DESCRIPTION
**Pull request by anthony-lipscomb-dev**
- [x] I've read, understood, and done my best to follow the [*CONTRIBUTING guidelines*](https://github.com/mindbody/conduit/blob/master/CONTRIBUTING.md).

This pull request includes (pick all that apply):

- [ ] Bugfixes
- [x] New features/Feature Gap
- [ ] Breaking changes
- [ ] Documentation updates
- [x] Unit tests
- [ ] Other

### Summary
This change will allow consuming applications of third party APIs to react accordingly when the user does not approve all requested scope permissions.

### Implementation
- Added `scope` property to the `OAuth2AuthorizationResponse` object as an array of Strings.
- Hooked this property up in the `OAuth2AuthorizationRedirectHandler`. The query string parameter `scope` is converted into a String array split by commas. This value is optional as users may choose to accept all requested permissions or none.

### Test Plan
- Added test to `OAuth2AuthorizationCodeTokenGrantStrategyTests` to verify the scope is not altered.
- Tested in live OAuth2SafariAuthorizationStrategy using Strava API, accepting and denying requested scope permissions.

### CODE REVIEW

**Comment by rminus** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Source files were modified – don’t forget to update the CHANGELOG if there are code changes.<br />
You can find examples within the <a href="https://github.com/mindbody/Conduit/blob/master/CHANGELOG.md">CHANGELOG.md</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by johnhammerlund** [comment]
@anthony-lipscomb-dev could you possibly add some context to this change? Strictly following OAuth 2, an Authorization Code redirect only includes the authorization code and state in the URL query or fragment (unless some service decides to add that custom behavior). The main exception to this are OpenID Connect Hybrid and Implicit flows, which are not yet directly supported.

Since it's ultimately up to each service how they might tweak the protocol spec to their own needs, we sometimes add flexibility based on the realistic development atmosphere. If this is a requirement, then I might suggest providing something more extendable and/or generic (i.e. a `customParameters` property, etc.)

**Comment by johnhammerlund** [comment]
Beyond that, I've been wanting to touch up the areas around Auth Code flow, so thanks for the PR! :+1:

**Comment by anthony-lipscomb-dev** [comment]
In the context of what I've needed it for, when the user Authorizes our application, there is a requested list of permissions that we send in the initial request that the user can either approve or deny so that the scope we initially request isn't what we are guaranteed.

eg.
https://www.example.com/oauth/authorize?state={{random}}&response_type=code&**scope=view_private,write**&redirect_uri={{URI}}&client_id={{id}}

would return data with the redirect header

{{redirect URI}}?state={{random but same}}&code={{token}}&**scope=view_private**
^^ missing write scope permission

I was using this as a reference https://www.oauth.com/oauth2-servers/access-tokens/access-token-response/

**Comment by johnhammerlund** [comment]
Not sure how this was closed, but I wasn't logged in when it says I did..

**Comment by johnhammerlund** [comment]
@anthony-lipscomb-dev that document refers to the code token exchange (authorization code token grant) response, not the authorization redirect.

The two parameters are listed at the bottom of the Authorization Code Response section here: https://www.oauth.com/oauth2-servers/authorization/the-authorization-response/
which are pulled straight from the OAuth 2.0 spec: https://tools.ietf.org/html/rfc6749#section-4.1.2

If Strava responds with scope in this authorization response, then that is custom behavior (not in spec for OAuth 2.0 or OpenID Connect). But, again, we rarely see services follow the spec to the tee, so there's definitely opportunity for supporting these cases more generically.

**Comment by anthony-lipscomb-dev** [comment]
@johnhammerlund Thank you for the documentation.

Looking through the RFC, section 4.2 - Implicit Grant. It shows the situation I am requesting to accommodate. 

This is not a Strava specific paradigm, but I have also found that Fitbit (https://dev.fitbit.com/build/reference/web-api/oauth2/) and I believe Facebook handles this the same way with `granted_scopes` as an optional parameter in the redirect, when implemented manually (https://developers.facebook.com/docs/facebook-login/manually-build-a-login-flow/). Being some of the larger implemented APIs, I'm sure I would be able to find others if I kept digging.

 What I did notice is that it is not always comma separated or listed as `scopes`, so a counter argument is that I would like to expose any of the remaining parameters to let the consuming application handle the specifics. That should be coming soon.

**Comment by eneko** [comment]
Much cleaner, I'd say, good updates.

There is a bit of repetition on passing `(code: code, state: state, queryItems: queryItemsDict)` since the first two are already contained in the third. I think it is easier to keep it as is for easier adoption on consuming apps and frameworks.

In regards of the tests, I'd like to see some tests to ensure the value is readable on the `queryItems` property, but not sure how to achieve that.

**Comment by anthony-lipscomb-dev** [comment]
@eneko I agree with it being repetitive. I also know that `code` is something that isn't optional in this workflow and state is, so was unsure how to approach it

**Comment by johnhammerlund** [comment]
@eneko @anthony-lipscomb-dev we could omit these fields and be a bit more specific with the property naming, i.e. `additionalParameters`, `customParameters`, etc. to imply included fields that fall outside of the spec or from hybrid / alternative protocols. What do you guys think?

@anthony-lipscomb-dev I think the real issue here is that `OAuth2AuthorizationResponse` does not yet support Implicit Grant (only Authorization Code is supported). I think this is a good step in that direction, but we eventually might want to make the response a generic container to support additional grant types.

**Comment by anthony-lipscomb-dev** [comment]
I am happy with this solution

File: Sources/Conduit/Auth/AuthorizationCodeGrant/OAuth2AuthorizationResponse.swift [hunk 1]
```diff
@@ -18,14 +18,19 @@ public struct OAuth2AuthorizationResponse {
     /// callback. This is primarily used to prevent CSRF attacks.
     public let state: String?
 
+    /// The acess the user has given to the consuming application.
+    /// This value is returned as a comma separated list in the query string.
+    public let customParameters: [String: String]?
```
Comment (johnhammerlund) [inline Sources/Conduit/Auth/AuthorizationCodeGrant/OAuth2AuthorizationResponse.swift]: Changing this to `[String: String]` and defaulting to an empty dictionary would simplify the structure & consuming callsites

Also, documentation needs to be updated

**Pull request merged**